cirq/study/resolver_test.py

```diff
@@ -159,11 +159,156 @@ def test_formulas_in_param_dict():
     r = cirq.ParamResolver({a: b + 1, b: 2, b + c: 101, 'd': 2 * e})
-    assert r.value_of('a') == 3
-    assert r.value_of('b') == 2
-    assert r.value_of(b + c) == 101
-    assert r.value_of('c') == c
-    assert r.value_of('d') == 2 * e
+    assert sympy.Eq(r.value_of('a'), 3)
+    assert sympy.Eq(r.value_of('b'), 2)
+    assert sympy.Eq(r.value_of(b + c), 101)
+    assert sympy.Eq(r.value_of('c'), c)
+    assert sympy.Eq(r.value_of('d'), 2 * e)
+
+
+def test_recursive_evaluation():
+    a = sympy.Symbol('a')
+    b = sympy.Symbol('b')
+    c = sympy.Symbol('c')
+    d = sympy.Symbol('d')
+    e = sympy.Symbol('e')
+    r = cirq.ParamResolver({
+        a: a,
+        b: e + 2,
+        c: b + d,
+        d: a + 3,
+        e: 0,
+    })
+
+    # sympy.Basic.subs evaluates in alphabetical order.
+    assert c.subs(r.param_dict) == b + a + 3
+
+    assert r.value_of(a) == a
+    assert sympy.Eq(r.value_of(b), 2)
+    assert sympy.Eq(r.value_of(c), a + 5)
+    assert sympy.Eq(r.value_of(d), a + 3)
+    assert sympy.Eq(r.value_of(e), 0)
+
+
+def test_unbound_recursion_halted():
+    a = sympy.Symbol('a')
+    b = sympy.Symbol('b')
+    c = sympy.Symbol('c')
+
+    # Non-recursive resolution ignores loops
+    r = cirq.ParamResolver({a: b, b: a})
+    assert r.value_of(a, recursive=False) == b
+    assert r.value_of(r.value_of(a, recursive=False), recursive=False) == a
+
+    # Self-definition is OK (this is a terminal symbol)
+    r = cirq.ParamResolver({a: a})
+    assert r.value_of(a) == a
+
+    r = cirq.ParamResolver({a: a + 1})
+    with pytest.raises(RecursionError):
+        _ = r.value_of(a)
+
+    r = cirq.ParamResolver({a: b, b: a})
+    with pytest.raises(RecursionError):
+        _ = r.value_of(a)
+
+    r = cirq.ParamResolver({a: b, b: c, c: b})
+    with pytest.raises(RecursionError):
+        _ = r.value_of(a)
+
+    r = cirq.ParamResolver({a: b + c, b: 1, c: a})
+    with pytest.raises(RecursionError):
+        _ = r.value_of(a)
+
+
+def test_resolve_unknown_type():
+    a = sympy.Symbol('a')
+    b = sympy.Symbol('b')
+    r = cirq.ParamResolver({a: b})
+    assert r.value_of(cirq.X) == cirq.X
+
+
+def test_compose():
+    """
+    Calling cirq.resolve_paramters on a ParamResolver composes that resolver
+    with the provided resolver.
+    """
+    a = sympy.Symbol('a')
+    b = sympy.Symbol('b')
+    c = sympy.Symbol('c')
+    d = sympy.Symbol('d')
+    r1 = cirq.ParamResolver({a: b})
+    r2 = cirq.ParamResolver({b: c + d})
+    r3 = cirq.ParamResolver({c: 12})
+
+    r12 = cirq.resolve_parameters(r1, r2)
+    assert r12.value_of('a') == c + d
+
+    r23 = cirq.resolve_parameters(r2, r3)
+    assert sympy.Eq(r23.value_of('b'), 12 + d)
+
+    r123 = cirq.resolve_parameters(r12, r3)
+    assert sympy.Eq(r123.value_of('a'), 12 + d)
+
+    r13 = cirq.resolve_parameters(r1, r3)
+    assert r13.value_of('a') == b
+
+
+@pytest.mark.parametrize('resolve_fn', [
+    cirq.resolve_parameters,
+    cirq.resolve_parameters_once,
+])
```

Comment:
Test parametrization is generally used to achieve separation of test data from test logic as in [table-driven tests](https://github.com/golang/go/wiki/TableDrivenTests). This is desirable because test logic is repetitive and we often want to perform the same operation on multiple inputs. Thus, this in a sense inverts the intended use of test parametrization: it uses test parameter to select code-under-test, but hardcodes test data within the test.

I suggest something like this instead:

```python
# Top of file up above
a, b, c, d, = sympy.Symbol('a'), sympy.Symbol('b'), ...

# ...

@pytest.mark.parametrize('p1, p2, p3', (
    ({a: b}, {c: d, d: 1}, {d: 2}),
    ({a: b}, {}, {b: 10}),
    ({a: 1}, {}, {}),
    ({}, {a: 1}, {}),
    ({}, {}, {a: 1}),
    # ...
))
def test_composition_is_associative(p1, p2, p3):
    r1 = cirq.ParamResolver(p1)
    r2 = cirq.ParamResolver(p2)
    r3 = cirq.ParamResolver(p3)
    for resolve_fn in (cirq.resolve_parameters, cirq.resolve_parameters_once):
        assert resolve_fn(r1, resolve_fn(r2, r3)) == resolve_fn(resolve_fn(r1, r2), r3)
```

This makes test logic general and cleanly separates the inputs we exercise within the decorator. Also, it makes it very easy to extend the test to cover new cases, e.g. if we find bugs.

Reply:
That would definitely be a lot cleaner! I think I'll parameterize on both, since that will make it easier to identify which case is failing by adding the function to the test name.

Reply:
Good point. You can probably use `itertools.product` here to make sure each test case is matched with each function :-)

Reply:
`pytest` has us covered: if a test has two separate `parametrize` lines, it will parameterize over their product 😃